// better-auth/dist/plugins/api-key/routes/list-api-keys.mjs
import { sessionMiddleware } from "../../../api/routes/session.mjs";
import "../../../api/index.mjs";
import { batchMigrateLegacyMetadata, listApiKeys, parseDoubleStringifiedMetadata } from "../adapter.mjs";
import { safeJSONParse } from "@better-auth/core/utils";
import { createAuthEndpoint } from "@better-auth/core/api";

//#region src/plugins/api-key/routes/list-api-keys.ts
function listApiKeys$1({ opts, schema, deleteAllExpiredApiKeys }) {
	return createAuthEndpoint("/api-key/list", {
		method: "GET",
		use: [sessionMiddleware],
		metadata: { openapi: {
			description: "List all API keys for the authenticated user",
			responses: { "200": {
				description: "API keys retrieved successfully",
				content: { "application/json": { schema: {
					type: "array",
					items: {
						type: "object",
						properties: {
							id: {
								type: "string",
								description: "ID"
							},
							name: {
								type: "string",
								nullable: true,
								description: "The name of the key"
							},
							start: {
								type: "string",
								nullable: true,
								description: "Shows the first few characters of the API key, including the prefix. This allows you to show those few characters in the UI to make it easier for users to identify the API key."
							},
							prefix: {
								type: "string",
								nullable: true,
								description: "The API Key prefix. Stored as plain text."
							},
							userId: {
								type: "string",
								description: "The owner of the user id"
							},
							refillInterval: {
								type: "number",
								nullable: true,
								description: "The interval in milliseconds between refills of the `remaining` count. Example: 3600000 // refill every hour (3600000ms = 1h)"
							},
							refillAmount: {
								type: "number",
								nullable: true,
								description: "The amount to refill"
							},
							lastRefillAt: {
								type: "string",
								format: "date-time",
								nullable: true,
								description: "The last refill date"
							},
							enabled: {
								type: "boolean",
								description: "Sets if key is enabled or disabled",
								default: true
							},
							rateLimitEnabled: {
								type: "boolean",
								description: "Whether the key has rate limiting enabled"
							},
							rateLimitTimeWindow: {
								type: "number",
								nullable: true,
								description: "The duration in milliseconds"
							},
							rateLimitMax: {
								type: "number",
								nullable: true,
								description: "Maximum amount of requests allowed within a window"
							},
							requestCount: {
								type: "number",
								description: "The number of requests made within the rate limit time window"
							},
							remaining: {
								type: "number",
								nullable: true,
								description: "Remaining requests (every time api key is used this should updated and should be updated on refill as well)"
							},
							lastRequest: {
								type: "string",
								format: "date-time",
								nullable: true,
								description: "When last request occurred"
							},
							expiresAt: {
								type: "string",
								format: "date-time",
								nullable: true,
								description: "Expiry date of a key"
							},
							createdAt: {
								type: "string",
								format: "date-time",
								description: "created at"
							},
							updatedAt: {
								type: "string",
								format: "date-time",
								description: "updated at"
							},
							metadata: {
								type: "object",
								nullable: true,
								additionalProperties: true,
								description: "Extra metadata about the apiKey"
							},
							permissions: {
								type: "string",
								nullable: true,
								description: "Permissions for the api key (stored as JSON string)"
							}
						},
						required: [
							"id",
							"userId",
							"enabled",
							"rateLimitEnabled",
							"requestCount",
							"createdAt",
							"updatedAt"
						]
					}
				} } }
			} }
		} }
	}, async (ctx) => {
		const session = ctx.context.session;
		const apiKeys = await listApiKeys(ctx, session.user.id, opts);
		deleteAllExpiredApiKeys(ctx.context);
		const returningApiKeys = apiKeys.map((apiKey) => {
			const { key: _key, ...rest } = apiKey;
			return {
				...rest,
				metadata: parseDoubleStringifiedMetadata(apiKey.metadata),
				permissions: rest.permissions ? safeJSONParse(rest.permissions) : null
			};
		});
		await ctx.context.runInBackgroundOrAwait(batchMigrateLegacyMetadata(ctx, apiKeys, opts));
		return ctx.json(returningApiKeys);
	});
}

//#endregion
export { listApiKeys$1 as listApiKeys };
